packages/drawer/src/ux-modal-service.ts

```diff
@@ -0,0 +1,277 @@
+    }
+    const element = document.createElement('ux-drawer');
+    element.setAttribute('dismiss.delegate', 'dismiss()');
+    element.setAttribute('ok.delegate', 'ok($event)');
+    if (options.position !== undefined) {
+      element.setAttribute('position', options.position);
+    }
+    if (options.moveToBodyTag === false) {
+      element.setAttribute('move-to-body-tag.bind', 'false');
+    }
+    if (options.overlayDismiss === false) {
+      element.setAttribute('overlay-dismiss', 'false');
+    }
+    if (options.keyboard !== undefined) {
+      bindingContext.keyboard = options.keyboard;
+      element.setAttribute('keyboard.bind', 'keyboard');
+    }
+    if (typeof options.restoreFocus === 'function') {
+      bindingContext.restoreFocus = options.restoreFocus;
+      element.setAttribute('restore-focus.bind', 'restoreFocus');
+    }
+    
+    element.setAttribute('host.bind', 'false');
+    
+    if (options.theme) {
+      bindingContext.theme = options.theme;
+      element.setAttribute('theme.bind', `theme`);
+    }
+    const compose = document.createElement('compose');
+    compose.setAttribute('view-model.ref', 'composeViewModel');
+    element.innerHTML = compose.outerHTML;
+    if (!options.host || options.host === 'body') {
+      options.host = document.body;
+    } else if (typeof options.host === 'string') {
+      options.host = document.querySelector(options.host) || document.body;
+    }
+    options.host.appendChild(element);
+    let childView = this.templatingEngine.enhance({ element: element, bindingContext: bindingContext });
+    
+    if (options.viewModel && bindingContext.composeViewModel) {
+      bindingContext.composeViewModel.viewModel = options.viewModel;
+    }
+    if (options.view && bindingContext.composeViewModel) {
+      bindingContext.composeViewModel.view = options.view;
+    }
+    if (options.model && bindingContext.composeViewModel) {
+      bindingContext.composeViewModel.model = options.model;
+    }
+
+    const controllers = (childView as any).controllers as Controller[];
+    
+    const drawer: UxDrawer =  controllers[0].viewModel as UxDrawer;
+    const drawerIndex = this.drawerIndex;
+    const whenClosed: Promise<ModalServiceResult> = new Promise((resolve) => {
+      this.eventAggregator.subscribeOnce(`drawer-${drawerIndex}-resolve`, (result: ModalServiceResult) => {
+        resolve(result);
+      });      
+    });
+
+    (drawer as UxDrawer & ModalServiceDrawer).whenClosed = () => {
+      return whenClosed;
+    }
+    bindingContext.dismiss = () => {
+      drawer.detached();
+      const parent = drawer.element.parentNode;
+      if (!parent) { return; }
+      parent.removeChild(drawer.element);
+      this.eventAggregator.publish(`drawer-${drawerIndex}-resolve`, {
+        wasCancelled: true,
+        output: null
+      });
+    }
+    bindingContext.ok = (event: CustomEvent) => {
+      drawer.detached();
```

Comment:
`detached` being called before the actual removal of the element is quite unexpected, compared with the rest. I think we can add another lifecycle here if needed, maybe `detaching`/`deactivating` @EisenbergEffect

Reply:
May I explain what is going on here.

When a drawer is dismissed or okeyed (could be inline be being removed from the DOM or from the ModalService), then it will do a few things.

1. It will try to call `canDeactivate` on its attached view-model (if any) (and stop here if returns false or Promise<false> or throw
2. It will call `detached` on its attached view-model (if any)
3. It will call `deactivate` on its attached view-model (if any)
4. It will do all the hiding tasks
5. Once this is completed, the `ux-drawer` dispatch a `CustomEvent` (either `ok` or `dismiss`)
6. Here we find the code above being a listener on the `ux-drawer`. It should be noted that these listeners are only added to `ux-drawer` elements that have been inserted by the ModalService (not inline). Because the ModalService has included the `ux-drawer` element in the DOM, it is now responsible for taking it away with the code below:

```ts
  bindingContext.dismiss = () => {
      drawer.detached();
      const parent = drawer.element.parentNode;
      if (!parent) {
        return;
      }
      parent.removeChild(drawer.element);
      this.eventAggregator.publish(`drawer-${drawerIndex}-resolve`, {
        wasCancelled: true,
        output: null
      });
    }
    bindingContext.ok = (event: CustomEvent) => {
      drawer.detached();
      const parent = drawer.element.parentNode;
      if (!parent) {
        return;
      }
      parent.removeChild(drawer.element);
      this.eventAggregator.publish(`drawer-${drawerIndex}-resolve`, {
        wasCancelled: false,
        output: event.detail
      });
    }
```

I've doing this be getting the `parent` and calling `removeChild`. However I discovered that `detached()` was not called on the `UxDrawer` VM in this particular case. So I thought to add it here manually before doing the actual remove. I did this _before_ the actual removal because I was afraid that doing it after might cause some errors.

Now I am wondering if my approach is correct. Maybe I should remove the `ux-drawer` differently than with a `removeChild()` ? Or should I call `detached()` later on ? Or else ?